examples/CPMPY_Pred+Opt.ipynb

```diff
@@ -0,0 +1,906 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "240\n"
+     ]
+    }
+   ],
+   "source": [
+    "weights = [data['weights'].tolist()]\n",
+    "weights = np.array(weights)\n",
+    "n_items = 48\n",
+    "capacity = 60\n",
+    "print(np.sum(weights))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### The weights are the costs of each item. The weights are kept same for each instance. The summation of 48 weights are 240. The knapsack capacity in this example is 60"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### We reshape the data in groups of 48 so that each group is an instance."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "(552, 48, 8) (237, 48, 8)\n"
+     ]
+    }
+   ],
+   "source": [
+    "x_train = x_train.reshape(-1,48,x_train.shape[1])\n",
+    "y_train = y_train.reshape(-1,48)\n",
+    "x_test = x_test.reshape(-1,48,x_test.shape[1])\n",
+    "y_test = y_test.reshape(-1,48)\n",
+    "print(x_train.shape,x_test.shape)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### As we see, we have 552 training instances and 237 test instances"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Next we wil randomize the training and test data"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "x = np.concatenate((x_train, x_test), axis=0)\n",
+    "y = np.concatenate((y_train,y_test), axis=0)\n",
+    "x,y = sklearn.utils.shuffle(x,y,random_state=0)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## This is the knpacak solver on CPMPY"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "def knapsack_solver(y, weights,capacity,n_items):\n",
+    "    x = boolvar(shape=n_items, name=\"x\")\n",
+    "\n",
+    "    model = Model(sum(x*weights[0]) <= capacity, maximize=sum(x*(1000*y).astype(int)))\n",
+    "    # We have to convert the y variable into integer, as ORtools only accepts integer variable\n",
+    "    model.solve()\n",
+    "    return x.value()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Now we create an wrapper object, which repeatedly solve the knapsack problem where the cost are fed as pytorch tensor"
+   ]
+  },
+  {
+   "cell_type": "code",
```

Comment:
what about combining this cell and the previous into something like:

```
class problem_solver():
    def __init__(self, weights, capacity):
        self.x = boolvar(shape=w.shape)
        self.m = CPM_ortools(Model( sum(x*w <= c) ))

    def solve_cost(cost):
        # convert to int for ortools
        cost = (1000*cost).astype(int)
        self.model.maximize((self.x*cost).sum())
        self.solve()
        return self.x.value()

   def solve_from_pytorch(tcost):
         # pytorch in pytorch out
         cost = tcost..detach().numpy()
         ret = solve_cost(cost)
         return torch.from_numpy(ret).float()

```

like this?